packages/@react-aria/focus/src/FocusScope.tsx

```diff
@@ -611,3 +620,106 @@ function last(walker: TreeWalker) {
   } while (last);
   return next;
 }
+
+
+class Tree {
+  private root: Node;
+  private fastMap = new Map<ScopeRef, Node>();
+
+  constructor() {
+    this.root = new Node({scopeRef: null});
+    this.fastMap.set(null, this.root);
+  }
+
+  get size() {
+    return this.fastMap.size;
+  }
+
+  getNode(data: ScopeRef) {
+    return this.fastMap.get(data);
+  }
+
+  addNode(scopeRef: ScopeRef, parent: ScopeRef, nodeToRestore?: FocusableElement) {
+    let parentNode = this.fastMap.get(parent ?? null);
+    let node = new Node({scopeRef});
+    parentNode.addChild(node);
+    node.parent = parentNode;
+    this.fastMap.set(scopeRef, node);
+    if (nodeToRestore) {
+      node.nodeToRestore = nodeToRestore;
+    }
+  }
+
+  removeNode(scopeRef: ScopeRef) {
+    // never remove the root
+    if (scopeRef === null) {
+      return;
+    }
+    let node = this.fastMap.get(scopeRef);
+    let parentNode = node.parent;
```

Comment:
Try this:

```suggestion
    let parentNode = node.parent;
    parentNode.children.forEach(
      child => {
        if (
          child !== node &&
          node.nodeToRestore &&
          child.nodeToRestore &&
          node.scopeRef.current &&
          isElementInScope(child.nodeToRestore, node.scopeRef.current)
        ) {
          child.nodeToRestore = node.nodeToRestore;
        }
      }
    );
```

Reply:
This does not work for me, did it work for you? I would be surprised if it did with the CRUD table, what is it solving for?
it might be easier to think through if it was written like
```
    let parentNode = node.parent;
    parentNode.children.forEach(
      sibling => {
        if (
          sibling !== node &&
          node.nodeToRestore &&
          sibling.nodeToRestore &&
          node.scopeRef.current &&
          isElementInScope(sibling.nodeToRestore, node.scopeRef.current)
        ) {
          sibling.nodeToRestore = node.nodeToRestore;
        }
      }
    );
```

Reply:
hmmm actually, that works in a live example, but not in unit tests